File: apps/mobile/src/models/Request.ts
import { Model } from '@nozbe/watermelondb';
import { field, date, readonly, relation } from '@nozbe/watermelondb/decorators';
import type { Associations } from '@nozbe/watermelondb/Model';
import type Family from './Family';
import type Wallet from './Wallet';

export type RequestStatus = 'draft' | 'pending' | 'approved' | 'rejected' | 'cancelled';

export default class Request extends Model {
  static table = 'requests';

  static associations: Associations = {
    families: { type: 'belongs_to', key: 'family_id' },
    wallets: { type: 'belongs_to', key: 'wallet_id' },
  };

  @field('family_id') familyId!: string;
  @field('requester_member_id') requesterMemberId!: string;
  @field('wallet_id') walletId!: string;
  @field('amount_cents') amountCents!: number;
  @field('category') category?: string;
  @field('notes') notes?: string;
  @field('attachment_url') attachmentUrl?: string;
  @field('status') status!: RequestStatus;
  @field('decision_by_member_id') decisionByMemberId?: string;
  @date('decision_at') decisionAt?: Date;
  @field('rejection_reason') rejectionReason?: string;

  @readonly @date('created_at') createdAt!: Date;
  @readonly @date('updated_at') updatedAt!: Date;

  @relation('families', 'family_id') family!: Family;
  @relation('wallets', 'wallet_id') wallet!: Wallet;
}

File: src-tauri/src/models/request.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum RequestStatus {
    Draft,
    Pending,
    Approved,
    Rejected,
    Cancelled,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum RequestCategory {
    Food,
    Transport,
    Education,
    Entertainment,
    Clothing,
    Health,
    Other,
}

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
pub struct Request {
    pub id: i64,
    pub uuid: String,
    pub family_id: i64,
    pub requester_member_id: i64,
    pub wallet_id: i64,
    pub amount_cents: i64,
    pub category: Option<String>,
    pub notes: Option<String>,
    pub attachment_url: Option<String>,
    pub status: String,
    pub decision_by_member_id: Option<i64>,
    pub decision_at: Option<String>,
    pub rejection_reason: Option<String>,
    pub sync_version: i64,
    pub last_synced_at: Option<String>,
    pub is_deleted: i64,
    pub created_at: String,
    pub updated_at: String,
}
